server/vmap/trillian_map_server.go

```diff
@@ -74,33 +78,47 @@ func (t *TrillianMapServer) GetLeaves(ctx context.Context, req *trillian.GetMapL
 
 	smtReader := merkle.NewSparseMerkleTreeReader(req.Revision, hasher, tx)
 
-	leaves, err := tx.Get(ctx, req.Revision, req.Index)
-	if err != nil {
-		return nil, err
-	}
-	glog.Infof("%v: wanted %v leaves, found %v", mapID, len(req.Index), len(leaves))
+	inclusions := make([]*trillian.MapLeafInclusion, 0, len(req.Index))
+	found := 0
+	for _, index := range req.Index {
+		// TODO: specify the index length in the tree specification.
+		if got, want := len(index), 32; got != want {
+			return nil, status.Errorf(codes.InvalidArgument,
+				"len(%x): %v, want %v", index, got, want)
+		}
+		proof, err := smtReader.InclusionProof(ctx, req.Revision, index)
+		if err != nil {
+			return nil, err
+		}
 
-	resp := &trillian.GetMapLeavesResponse{
-		MapLeafInclusion: make([]*trillian.MapLeafInclusion, len(leaves)),
-	}
-	for i, leaf := range leaves {
-		proof, err := smtReader.InclusionProof(ctx, req.Revision, leaf.Index)
+		// Create placeholder leaf for non-existent leaves.
+		leaf := &trillian.MapLeaf{
+			Index: index,
+		}
+
+		// Fetch the leaf if it exists.
+		leaves, err := tx.Get(ctx, req.Revision, [][]byte{index})
 		if err != nil {
 			return nil, err
 		}
-		// Copy the leaf from the iterator, which gets overwritten
-		value := leaf
-		resp.MapLeafInclusion[i] = &trillian.MapLeafInclusion{
-			Leaf:      &value,
-			Inclusion: proof,
+		if len(leaves) == 1 {
+			leaf = &leaves[0]
+			found++
```

Comment:
I don't really know the map stuff, so I found this a bit hard to follow (both before and after!).

Should (exists? leaf) == (exists? proof) ? If so, it might be nice to check for internal consistency (which I guess is `(len(leaves) == 0)` iff `(proof == nil)`)